src/Workflow/MoveJatsAssets.php

```diff
@@ -0,0 +1,164 @@
+            $this->concurrency
+        )->wait();
+    }
+
+    /**
+     * @return iterable<Element>
+     */
+    private function findLinkedElements(Document $document) : iterable
+    {
+        $document->registerNamespace('jats', 'http://jats.nlm.nih.gov');
+        $document->registerNamespace('xlink', 'http://www.w3.org/1999/xlink');
+
+        yield from $document('//jats:article//jats:*[@xlink:href]');
+    }
+
+    private function processAssets(iterable $elements, PutTask $task) : iterable
+    {
+        foreach ($elements as $element) {
+            if ($this->shouldProcess($element)) {
+                yield $this->processAsset($element, $task);
+            }
+        }
+    }
+
+    private function processAsset(Element $asset, PutTask $task) : PromiseInterface
+    {
+        return new Coroutine(
+            function () use ($asset, $task) : iterable {
+                /** @var ResponseInterface $response */
+                $response = yield $this->client->requestAsync('GET', element_uri($asset));
+
+                $contentType = parse_media_type($response->getHeaderLine('Content-Type'));
+                /** @var resource $stream */
+                $stream = $response->getBody()->detach();
+                $path = $this->pathFor($task, $contentType, $stream);
+
+                $this->updateElement($asset, $contentType, $path);
+                $this->deployAsset($contentType, $stream, $path);
+            }
+        );
+    }
+
+    /**
+     * @param resource $stream
+     */
+    private function deployAsset(array $contentType, $stream, string $path) : void
+    {
+        $this->filesystem->putStream(
+            $path,
+            $stream,
+            [
+                'mimetype' => implode('/', $contentType),
+                'visibility' => AdapterInterface::VISIBILITY_PUBLIC,
+            ]
+        );
+    }
+
+    private function updateElement(Element $element, array $contentType, string $path) : void
+    {
+        $element->setAttribute('xlink:href', sprintf('%s/%s', $this->publicUri, $path));
+
+        if (in_array($element->localName, self::HAS_MIMETYPE_ATTRIBUTE, true)) {
+            $element->setAttribute('mimetype', $contentType[0]);
+            $element->setAttribute('mime-subtype', $contentType[1]);
+        }
+    }
+
+    private function shouldProcess(Element $element) : bool
+    {
+        $uri = element_uri($element);
+
+        return Uri::isAbsolute($uri) && preg_match($this->origin, (string) $uri);
+    }
+
+    /**
+     * @param resource $stream
+     */
+    private function pathFor(PutTask $task, array $contentType, $stream) : string
+    {
+        $hash = stream_hash($stream);
+
+        $path = sprintf('%s/v%s/%s', $task->getItemId(), $task->getItemVersion(), $hash);
+
+        $extension = ExtensionGuesser::getInstance()->guess(implode('/', $contentType));
+        if ($extension) {
+            $path .= ".{$extension}";
+        }
+
+        return $path;
```

Comment:
weird case, but if there are identical images they end up on the same path. This works as it just de-duplicates them, but any possible race conditions between different concurrent uploads?
S3 as a backend should behave as last-write-wins (eventually) and all the writes are the same content.

Reply:
It is a bit of an edge case. Same with identical images existing at two different URIs (produce the same result). It's redundant work that could be resolved somehow, but ignoring for the moment as is _probably_ safe.